src/NexusMods.Telemetry/EventSender.cs

```diff
@@ -0,0 +1,253 @@
+
+    private ulong _lastGeneratedId;
+    private ulong _highestSeenId;
+
+    /// <summary>
+    /// Inserts the event into the queue.
+    /// </summary>
+    public void AddEvent(EventDefinition definition, EventMetadata metadata)
+    {
+        // NOTE(erri120): Since these are just tracking events, it's okay if we
+        // loose some of them. As such, a simple "ring buffer" represented by an
+        // array is completely fine and this shouldn't impact performance for
+        // any consumer inserting events.
+        Debug.Assert(metadata.IsValid());
+
+        var id = Interlocked.Increment(ref _lastGeneratedId);
+        var newIndex = (int)(id % Limit);
+        Debug.Assert(newIndex >= 0);
+        Debug.Assert(newIndex < _insertRingBuffer.Length);
+
+        var tuple = new ValueTuple<EventDefinition, EventMetadata, ulong>(definition, metadata, id);
+        _insertRingBuffer[newIndex] = tuple;
+    }
+
+    public async ValueTask Run()
+    {
+        PrepareArrays();
+
+        var span = PrepareSpan(_sortedReadingCopy, _highestSeenId);
+        if (span.IsEmpty) return;
+
+        _highestSeenId = span[^1].Item3;
+
+        var userId = _loginManager.UserInfo?.UserId ?? Optional<UserId>.None;
+
+        try
+        {
+            PrepareRequest(_writer, span, userId);
+            await SendEvents(_writer.WrittenMemory, _httpClient, CancellationToken.None);
+        }
+        finally
+        {
+            _writer.ResetWrittenCount();
+        }
+    }
+
+    /// <summary>
+    /// Prepares the HTTP request data.
+    /// </summary>
+    private static void PrepareRequest(IBufferWriter<byte> writer, ReadOnlySpan<ValueTuple<EventDefinition, EventMetadata, ulong>> events, Optional<UserId> userId)
+    {
+        // https://developer.matomo.org/api-reference/tracking-api#bulk-tracking
+        using (var sb = ZString.CreateUtf8StringBuilder(notNested: true))
+        {
+            sb.Append("{ \"requests\": [");
+            sb.CopyTo(writer);
+        }
+
+        var isFirst = true;
+        foreach (var eventTuple in events)
+        {
+            if (!isFirst)
+            {
+                using var sb = ZString.CreateUtf8StringBuilder(notNested: true);
+                sb.Append(",");
+                sb.CopyTo(writer);
+            }
+
+            isFirst = false;
+            SerializeEvent(writer, eventTuple, userId);
+        }
+
+        using (var sb = ZString.CreateUtf8StringBuilder(notNested: true))
+        {
+            sb.Append("] }");
+            sb.CopyTo(writer);
+        }
+    }
+
+    /// <summary>
+    /// Serializes the event into the buffer as an UTF8 encoded string.
+    /// </summary>
+    [SuppressMessage("ReSharper", "StringLiteralTypo")]
+    private static void SerializeEvent(IBufferWriter<byte> writer, ValueTuple<EventDefinition, EventMetadata, ulong> tuple, Optional<UserId> userId)
```

Comment:
Should probably serialize to passed StringBuilder instead of directly onto the buffer

Reply:
See above.